File: java edition/rotateString.java
import java.util.*;

public class rotateString{
    public static void main(String[] args){
        char[] str = new char[]{'a', 'b', 'c', 'd', 'e', 'f', 'g'};
        rotateString(str, 3);

    }

    public static void rotateString(char[] str, int offset) {
        if(str == null || str.length == 0 || offset == 0){
            return;
        }
        offset = offset % str.length;
        reverse(str, 0, offset);
        reverse(str, offset + 1, str.length - 1);
        reverse(str, 0, str.length - 1);
    }

    private static void reverse(char[] str, int start, int end){
        while(start < end){
            System.out.println(start + "\t" + end);
            char tmp = str[start];
            str[start] = str[end];
            str[end] = tmp;
            start += 1;
            end -= 1;
        }
    }
}

File: rotateString.py
class Solution:
    # @param s: a list of char
    # @param offset: an integer
    # @return: nothing
    def rotateString(self, s, offset):
        if not s:
            return
        offset = offset % len(s)

        self.reverse(s, 0, offset)
        self.reverse(s, offset + 1, len(s) - 1)
        self.reverse(s, 0, len(s) - 1)
        print s

    def reverse(self, s, start, end):
        while start < end:
            tmp = s[start]
            s[start] = s[end]
            s[end] = tmp
            start += 1
            end += 1

if __name__ == '__main__':
    a = Solution()
    a.rotateString("abcdefg", 3)
